File: WaterlooBankObjC/WaterlooBankObjC/ViewController.m
//
//  ViewController.m
//  WaterlooBankObjC
//
//  Created by Jonathan Ellis on 05/07/2015.
//  Copyright (c) 2015 iProov Ltd. All rights reserved.
//

#import "ViewController.h"
#import "MainViewController.h"
#import <iProov/iProov-Swift.h>

#define WB_SERVICE_PROVIDER @"a73e90cf90e3ede70fd38639ef621f072aca8364"

@interface ViewController ()

@property (strong, nonatomic) IBOutlet UITextField *usernameTextField;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.title = @"Waterloo Bank (ObjC)";
    self.navigationItem.backBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"Logout" style:UIBarButtonItemStylePlain target:nil action: nil];

}

- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    [textField resignFirstResponder];
    
    return YES;
}

- (IBAction)loginButtonPressed:(UIButton *)sender {
    
    if (self.usernameTextField.text.length == 0) {
        
        UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Error" message:@"Please enter your username/email to login." preferredStyle:UIAlertControllerStyleAlert];
        [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil]];
        [self presentViewController:alert animated:YES completion:nil];
        
        return;
    }
    
    [self login:self.usernameTextField.text];

}

- (void)login:(NSString *)username {
    
    [IProov verifyWithServiceProvider:WB_SERVICE_PROVIDER
                             username:username
                             animated:YES
                              success:^(NSString * _Nonnull token) {
                                  [self performSegueWithIdentifier:@"ShowMain" sender:token];
                                  self.usernameTextField.text = nil;
                              
                              } failure:^(NSString * _Nonnull reason) {
                                  UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Login failed" message:reason preferredStyle:UIAlertControllerStyleAlert];
                                  [alert addAction:[UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:nil]];
                                  [alert addAction:[UIAlertAction actionWithTitle:@"Retry" style:UIAlertActionStyleDefault handler:^(UIAlertAction *action) {
                                      [self login:username];
                                  }]];
                                  
                                  [self presentViewController:alert animated:YES completion:nil];
                              
                              } error:^(NSError * _Nonnull error) {
                                  
                                  if (error.code == IProovErrorCodeUserPressedBack || error.code == IProovErrorCodeUserPressedHome) { // Ignore cases where the user voluntarily pressed the back button
                                      return;
                                  }
                                  
                                  UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Login failed" message:error.localizedDescription preferredStyle:UIAlertControllerStyleAlert];
                                  [alert addAction:[UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:nil]];
                                  [alert addAction:[UIAlertAction actionWithTitle:@"Retry" style:UIAlertActionStyleDefault handler:^(UIAlertAction *action) {
                                      [self login:username];
                                  }]];
                                  [self presentViewController:alert animated:YES completion:nil];
                              }];
}

- (IBAction)registerButtonPressed:(UIButton *)sender {
    
    if (self.usernameTextField.text.length == 0) {
        
        UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Error" message:@"Please enter your username/email to register." preferredStyle:UIAlertControllerStyleAlert];
        [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil]];
        [self presentViewController:alert animated:YES completion:nil];
        
        return;
    }
    
    [self registerUser:self.usernameTextField.text];
    
}

- (void)registerUser:(NSString *)username {

    [IProov enrolWithServiceProvider:WB_SERVICE_PROVIDER
                            username:username
                            animated:YES
                             success:^(NSString * _Nonnull token) {
                                 [self performSegueWithIdentifier:@"ShowMain" sender:token];
                                 self.usernameTextField.text = nil;
                                                 
                            } failure:^(NSString * _Nonnull reason) {
                                UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Registration failed" message:reason preferredStyle:UIAlertControllerStyleAlert];
                                [alert addAction:[UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:nil]];
                                [alert addAction:[UIAlertAction actionWithTitle:@"Retry" style:UIAlertActionStyleDefault handler:^(UIAlertAction *action) {
                                    [self registerUser:username];
                                }]];
                                                 
                                [self presentViewController:alert animated:YES completion:nil];
                                                 
                            } error:^(NSError * _Nonnull error) {
                                                 
                                 if (error.code == IProovErrorCodeUserPressedBack || error.code == IProovErrorCodeUserPressedHome) { // Ignore cases where the user voluntarily pressed the back button
                                     return;
                                 }
                                 
                                 UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Registration failed" message:error.localizedDescription preferredStyle:UIAlertControllerStyleAlert];
                                 [alert addAction:[UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:nil]];
                                 [alert addAction:[UIAlertAction actionWithTitle:@"Retry" style:UIAlertActionStyleDefault handler:^(UIAlertAction *action) {
                                     [self registerUser:username];
                                 }]];
                                 [self presentViewController:alert animated:YES completion:nil];
                             }];
    
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    
    if ([segue.identifier isEqualToString:@"ShowMain"]) {
        
        MainViewController *mainViewController = segue.destinationViewController;
        
        mainViewController.token = sender;
    }
    
}

@end

File: WaterlooBank/WaterlooBank/ViewController.swift
//
//  ViewController.swift
//  WaterlooBank
//
//  Created by Jonathan Ellis on 22/06/2015.
//  Copyright (c) 2015 iProov Ltd. All rights reserved.
//

import UIKit
import iProov

class ViewController: UIViewController, UITextFieldDelegate {
    
    let serviceProvider: String? = nil //TODO: define your API key here

    @IBOutlet var usernameTextField: UITextField!

    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.title = "Waterloo Bank"
        self.navigationItem.backBarButtonItem = UIBarButtonItem(title: "Logout", style: .plain, target: nil, action: nil)
    }
    
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        textField.resignFirstResponder()
        return true
    }
    
    @IBAction func loginButtonPressed(_ sender: UIButton) {
        
        if usernameTextField.text?.isEmpty != false {
            let alert = UIAlertController(title: "Error", message: "Please enter your username/email to login.", preferredStyle: .alert)
            alert.addAction(UIAlertAction(title: "OK", style: .default, handler: nil))
            self.present(alert, animated: true, completion: nil)
            
            return
        }
        
        login(usernameTextField.text!)
    }
    
    func login(_ username: String) {
        
        guard serviceProvider != nil else{
            let alert = UIAlertController(title: "Missing API key", message: "Please edit serviceProvider in ViewController.swift", preferredStyle: .alert)
            alert.addAction(UIAlertAction(title: "OK", style: .default, handler: nil))
            self.present(alert, animated: true, completion: nil)
            
            return
        }
        
        IProov.verify(withServiceProvider: serviceProvider!, username: username, animated: true) { (result) in
            
            switch result {
            case let .success(token):
                self.performSegue(withIdentifier: "ShowMain", sender: token)
                self.usernameTextField.text = nil
                
            case let .failure(reason):
                let alert = UIAlertController(title: "Login failed", message: reason.reason, preferredStyle: .alert)
                alert.addAction(UIAlertAction(title: "Cancel", style: .cancel, handler: nil))
                alert.addAction(UIAlertAction(title: "Retry", style: .default, handler: { (action) -> Void in
                    self.login(username)
                }))
                self.present(alert, animated: true, completion: nil)
                
            case let .error(error):
                let alert = UIAlertController(title: "Login failed", message: error.localizedDescription, preferredStyle: .alert)
                alert.addAction(UIAlertAction(title: "Cancel", style: .cancel, handler: nil))
                alert.addAction(UIAlertAction(title: "Retry", style: .default, handler: { (action) -> Void in
                    self.login(username)
                }))
                self.present(alert, animated: true, completion: nil)
                
            default: break

            }
            
        }

    }


    @IBAction func registerButtonPressed(_ sender: UIButton) {
        
        if usernameTextField.text?.isEmpty != false {
            let alert = UIAlertController(title: "Error", message: "Please enter your username/email to register.", preferredStyle: .alert)
            alert.addAction(UIAlertAction(title: "OK", style: .default, handler: nil))
            self.present(alert, animated: true, completion: nil)
            
            return
        }
        
        register(usernameTextField.text!)
    }
    
    
    func register(_ username: String) {
        
        guard serviceProvider != nil else{
            let alert = UIAlertController(title: "Missing API key", message: "Please edit serviceProvider in ViewController.swift", preferredStyle: .alert)
            alert.addAction(UIAlertAction(title: "OK", style: .default, handler: nil))
            self.present(alert, animated: true, completion: nil)
            
            return
        }
        
        IProov.enrol(withServiceProvider: serviceProvider!, username: username, animated: true) { (result) in
            
            switch result {
                
            case let .success(token):
                self.performSegue(withIdentifier: "ShowMain", sender: token)
                self.usernameTextField.text = nil
                
            case let .failure(reason):
                
                let alert = UIAlertController(title: "Registration failed", message: reason.reason, preferredStyle: .alert)
                alert.addAction(UIAlertAction(title: "Cancel", style: .cancel, handler: nil))
                alert.addAction(UIAlertAction(title: "Retry", style: .default, handler: { (action) -> Void in
                    self.register(username)
                }))
                self.present(alert, animated: true, completion: nil)
                
            case let .error(error):
                
                let alert = UIAlertController(title: "Registration failed", message: error.localizedDescription, preferredStyle: .alert)
                alert.addAction(UIAlertAction(title: "Cancel", style: .cancel, handler: nil))
                alert.addAction(UIAlertAction(title: "Retry", style: .default, handler: { (action) -> Void in
                    self.register(username)
                }))
                self.present(alert, animated: true, completion: nil)
                
            default: break
                
            }
            
        }
        
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if segue.identifier == "ShowMain" {
            let mainViewController = segue.destination as! MainViewController
            mainViewController.token = sender as! String
        }
    }
}

